[1108] MP Rest Client 2.0 - First Look [staging]

andymc12 opened the pull request:
Part of #1108

jakub-pomykala commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+Most servers will handle that just fine, but other servers will require that the HTTP request be a single key with a 
+comma-separated list of values, like: `?myParam=a,b,c`
+
+Still other servers will require array-like syntax, such as: `?myParam[]=a&myParam[]=b&myParam[]=c`
```
```suggestion
Other servers will still require array-like syntax, such as: `?myParam[]=a&myParam[]=b&myParam[]=c`
```

andymc12 replied:
Sounds good.

jakub-pomykala commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+[#proxy-server]
+== Proxy server support
+
+Many RESTful endpoints might require the use of a proxy server to access. MicroProfile Rest Client 2.0 makes it easier
```
I think we should either have the first "Microprofile" expanded and then all others as "MP" or just stick to "MicroProfile" right now sometimes it's "MP" and sometimes it's "MicroProfile"

andymc12 replied:
Agreed - I will change these to be "MP" after the first time it is referenced in the text.  I may leave it as "MicroProfile" in the closing summary though - let me know if you have a strong opinion against that.

jakub-pomykala replied:
Really good idea!

jakub-pomykala commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+== Support for Server Sent Events (SSEs)
+
+link:https://openliberty.io/guides/reactive-messaging-sse.html[Server Sent Events] are part of the HTML 5 spec and
+enables a server to push data to a client asynchronously via events - all over HTTP. The JAX-RS 2.1 spec enabled SSE
```
Is `JAX-RS 2.1` not supposed to be in backticks?

andymc12 replied:
In this case, I don't think it should be in backticks.  I'm referring to the specification/project, not the Liberty feature (which would go in backticks, like `jaxrs-2.1`).

tevans78 commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+
+
+[#query-param-style]
+== Use QueryParamStyle for specifying how collections of query parameters should be formatted
```
```suggestion
== Using QueryParamStyle to specify how collections of query parameters should be formatted
```

andymc12 replied:
Done

tevans78 commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+
+Alternatively the query parameter style can be declared through MP Config using a property like:
+
+`com.mypkg.MyMultiValuedQueryParamClient/mp-rest/queryParamStyle=ARRAY_PAIRS`
```
@andymc12 should we have a short table that lists the three options and gives an example of each. As it is you kinda have to infer which one is which?

tevans78 commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+
+Alternatively the query parameter style can be declared through MP Config using a property like:
+
+`com.mypkg.MyMultiValuedQueryParamClient/mp-rest/queryParamStyle=ARRAY_PAIRS`
```
Also, `MyMultiValuedQueryParamClient` is a rather long class name. The other sections below just use `MyClient`, that might be better here too?

andymc12 replied:
Sure - switched to `MyClient`

tevans78 commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+`com.mypkg.MyClient/mp-rest/proxyAddress=myproxy.xyz.com:1080`
+
+Note that this approach to setting the proxy server host and port should be preferred to using vendor-specific
+properties like `com.ibm.ws.jaxrs.client.proxy.host` and `com.ibm.ws.jaxrs.client.proxy.port`, though these properties
```
```suggestion
properties such as `com.ibm.ws.jaxrs.client.proxy.host` and `com.ibm.ws.jaxrs.client.proxy.port`, though these properties
```

tevans78 commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+Note that this approach to setting the proxy server host and port should be preferred to using vendor-specific
+properties like `com.ibm.ws.jaxrs.client.proxy.host` and `com.ibm.ws.jaxrs.client.proxy.port`, though these properties
+will still work. For proxy authentication, you can still use the `com.ibm.ws.jaxrs.client.proxy.username` and
+`com.ibm.ws.jaxrs.client.proxy.password` properties like:
```
```suggestion
`com.ibm.ws.jaxrs.client.proxy.password` properties. For example:
```

tevans78 commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+[#auto-redirects]
+== Automatically following redirects
+
+If a RESTful resource has relocated, often the HTTP response code will be in the 300 range and will indicate the new
```
```suggestion
If a RESTful resource has been relocated, often the HTTP response code will be in the 300 range and will indicate the new
```

tevans78 commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+
+First, the method (or interface) must be annotated with `@Produces(MediaType.SERVER_SENT_EVENTS)` to indicate that it
+expects the server to produce SSEs. Next, the method's return type must be `org.reactivestreams.Publisher`. The generic
+type can be `javax.ws.rs.sse.InboundSseEvent` from JAX-RS or it can be a primitive or complex object - the latter only
+works if (1) the server only sends one type of events (e.g. only sends `WeatherEvents` - then `Publisher<WeatherEvent>`
+would be applicable) and (2) if you have an entity provider that is capable of converting the events into the generic
+type (e.g. `MessageBodyReader<WeatherEvent>`). For most cases, if the remote server sends events using JSON, you can
+enable the `jsonb-1.0` feature in your Liberty server which will automatically register a JSON-B-based entity provider.
```
```suggestion
type can be `javax.ws.rs.sse.InboundSseEvent` (from JAX-RS), a primitive, a String or a complex type. Complex types may only be used if;
 1. the server only sends one type of event (e.g. only sends `WeatherEvents` - then `Publisher<WeatherEvent>` would be applicable)
_and_
 2. there is a registered entity provider capable of converting the events (e.g. `MessageBodyReader<WeatherEvent>`).
In most cases, if the remote server sends events using JSON, you can
enable the `jsonb-1.0` feature in your Liberty server which will automatically register a JSON-B-based entity provider.
```

tevans78 commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,201 @@
+
+== Summary
+
+MicroProfile Rest Client 2.0 has some powerful new features useful for building cloud native applications. There are
+still more features not covered here, but have been in Open Liberty's previous Rest Client support, including using CDI-
+enabled providers and supporting `ContextResolver` to customize the built-in JSON-B provider. These features and more
+are available on the link:https://github.com/eclipse/microprofile-rest-client/releases/tag/2.0[MP Rest Client 2.0 release page].
```
Does "JSON-B ContextResolver Support" not deserve it's own section? If not then it seems a little confusing to mention it here.
```suggestion
MicroProfile Rest Client 2.0 has some powerful new features useful for building cloud native applications. You can read more about these updates on the link:https://github.com/eclipse/microprofile-rest-client/releases/tag/2.0[MP Rest Client 2.0 release page].
```

andymc12 commented:
Pulled in changes from `draft` that addressed @tevans78 's review comments.  Tom and @jakub-pomykala, can you re-review and merge?  Thanks!

jakub-pomykala reviewed (approved):
LGTM

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+The latest release of MicroProfile's type-safe REST client has a lot of new and exciting features. In this post, we'll
+take a look at some of the new features and how you can use them in Open Liberty, including:
+
+. <<query-param-style,Using QueryParamStyle to specify how collections of query parameters should be formatted>>
```
I think this list should be unordered bullets rather than ordered steps, since it doesn't list sequential, task based info

andymc12 replied:
Agreed - will fix

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+
+
+[#query-param-style]
+== Using QueryParamStyle to specify how collections of query parameters should be formatted
```
suggested edit:

Using QueryParamStyle to specify how to format collections of query parameters 

If you agree, reminder to change the heading in the introduction list as well

andymc12 replied:
I think this should remain as it is.  I don't like the passive voice or the extra words either - but I think it better implies that the implementation (Liberty) handles the formatting - the user just tells us how to format it with the `QueryParamStyle`.

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+    String multiValues(@QueryParam("myParam") List<String> values);
+}
+----
+By default, if a caller invokes the the client with `multiValues(Arrays.asList("a", "b", "c"))` the MP Rest Client will
```
remove extra "the" 

if a caller invokes the the client

and add a period after 

`?myParam=a&myParam=b&myParam=c`

andymc12 replied:
Good catch! will fix

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+By default, if a caller invokes the the client with `multiValues(Arrays.asList("a", "b", "c"))` the MP Rest Client will
+produce an HTTP request with multiple key/value pairs - something like: `?myParam=a&myParam=b&myParam=c`
+
+Most servers will handle that just fine, but other servers will require that the HTTP request be a single key with a 
```
suggested edit

Most servers will handle that just fine, but other servers will require that the HTTP request be a single key ...

--->

Although most servers will handle that just fine, some servers might require the HTTP request to be a single key...

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+                                   .build(MyClient.class);
+----
+
+Alternatively the query parameter style can be declared through MP Config using a property like:
```
Alternatively, you can declare the query parameter style through MP Config by using a property like:

dmuelle replied:
Also, is it helpful to link to MP Config guide on first mention of MP Config?
In case the reader is not familiar...

https://www.openliberty.io/guides/microprofile-config.html or https://www.openliberty.io/guides/microprofile-config-intro.html

andymc12 replied:
Good idea - will do!

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+`com.mypkg.MyClient/mp-rest/queryParamStyle=ARRAY_PAIRS`
+
+.QueryParamStyle Options
+|=======================
```
I think this table needs an intro sentence -does it apply specifically to MP config or just generally to Rest Client?

Something like

"The following table lists the enum values that are available for xxxxx and provides an example oof the corresponding output for each value"

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+[#proxy-server]
+== Proxy server support
+
+Many RESTful endpoints might require the use of a proxy server to access. MP Rest Client 2.0 makes it easier
```
suggested edit
Many RESTful endpoints might require the use of a proxy server to access.

--->

You might need to use a proxy server to access some RESTful endpoints.

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+Many RESTful endpoints might require the use of a proxy server to access. MP Rest Client 2.0 makes it easier
+and more portable to specify a proxy server with the new `proxyAddress(host, port)` method on the `RestClientBuilder`
+class. For example, suppose you need to access an endpoint via a proxy server at myproxy.xyz.com on port 1080. You could
+build your rest client instance with code like:
```
should these values be monospaced?

suppose you need to access an endpoint via a proxy server at `myproxy.xyz.com` on port `1080`.

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+
+`com.mypkg.MyClient/mp-rest/proxyAddress=myproxy.xyz.com:1080`
+
+Note that this approach to setting the proxy server host and port should be preferred to using vendor-specific
```
should be preferred
--->
is preferred.

Also, is it possible to concisely state why it's better/preferred to use MP config than the vendor specific properties?

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+If a RESTful resource has been relocated, often the HTTP response code will be in the 300 range and will indicate the
+new location. Rather than handling the 3XX response and manually issuing a new request, MP Rest Client 2.0 allows rest
+client instances to automatically follow redirects. This can be configured programmatically while building the client
+instance or via MP Config. Here is an example of configuring auto-redirect via the `RestClientBuilder` API:
```
This can be configured programmatically while building the client instance or via MP Config.

--->

You can configure a client to automatically follow redirects either programmatically, when you build the client instance, or via MP Config.

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+== Support for Server Sent Events (SSEs)
+
+link:https://openliberty.io/guides/reactive-messaging-sse.html[Server Sent Events] are part of the HTML 5 spec and
+enables a server to push data to a client asynchronously via events - all over HTTP. The JAX-RS 2.1 spec enabled SSE
```
[Server Sent Events] are part of the HTML 5 spec and enables a server to push data to a client asynchronously via events - all over HTTP.

--->

[Server Sent Events], part of the HTML 5 spec, enable a server to push data to a client asynchronously via events,  over HTTP.

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+support for both the client and server. Now you can consume SSE events from the type-safe MP Rest Client.
+
+The MP Rest Client specification uses the link:http://www.reactive-streams.org/[Reactive Streams] APIs to consume
+events. A client interface capable of consuming SSEs would look something like this:
```
A client interface capable of consuming SSEs would look something like this:

--->

A client interface capable of consuming SSEs looks something like this:

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+
+First, the method (or interface) must be annotated with `@Produces(MediaType.SERVER_SENT_EVENTS)` to indicate that it
+expects the server to produce SSEs. Next, the method's return type must be `org.reactivestreams.Publisher`. The generic
+type can be `javax.ws.rs.sse.InboundSseEvent` (from JAX-RS), a primitive, a String or a complex type. Complex types may only be used if;
```
use serial comma:

The generic type can be `javax.ws.rs.sse.InboundSseEvent` (from JAX-RS), a primitive, a String, or a complex type.
 
Also, avoid "may"

Complex types can only be used if....

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+
+2) there is a registered entity provider capable of converting the events (e.g. `MessageBodyReader<WeatherEvent>`).
+
+In most cases, if the remote server sends events using JSON, you can enable the `jsonb-1.0` feature in your Liberty
```
add comma:

you can enable the jsonb-1.0 feature in your Liberty server, which automatically registers a JSON-B-based entity provider.

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+server which will automatically register a JSON-B-based entity provider.
+
+
+Once you've invoked one of these methods, you should be able to register one or more `Subscriber` instances to the
```
avoid past perfect and subjunctive:

Once you invoke one of these methods, you can register one or more `Subscriber` instances...

dmuelle commented on posts/2021-03-24-whats-new-in-MP-Rest-Client2.0.adoc:
```diff
@@ -0,0 +1,212 @@
+
+== Summary
+
+MicroProfile Rest Client 2.0 has some powerful new features useful for building cloud native applications. You can read
```
MicroProfile Rest Client 2.0 has some powerful new features useful for building cloud native applications.
--->
MicroProfile Rest Client 2.0 has some powerful new features that are useful for building cloud native applications.

dmuelle reviewed (commented):
ID review- see my comments in line. Let me know if any questions or if I've misinterpreted anything. Thanks

Also- reminder that any edits that are made here should also be delivered to draft, preferably before staging

andymc12 commented:
Thanks for the review @dmuelle - I've opened PR #1252 to make your suggested changes.  Would you mind reviewing it?  Thanks again!

andymc12 commented:
Ok - I think I've addressed all the issues from the ID review and cherry-picked those changes from draft/staging to this PR (for prod).  @dmuelle @mbroz2 @tevans78 @jakub-pomykala - do any of you want to take a final look at it?  After that, could one of you merge?  Thanks again for all the review and feedback!

dmuelle commented:
Looks good @andymc12 . We typically merge staging->prod to publish a post. Did you cherry-pick because with the flurry of MP blog activity, there might be things in staging that are not scheduled to publish yet?

mbroz2 commented:
like @dmuelle mentioned, there should be no need to cherry-pick into prod.  We can just merge staging into prod:
https://github.com/OpenLiberty/blogs/pull/1256

gljrobinson commented:
@dmuelle should this PR still be open if the blog is already published?

mbroz2 commented:
Going to close this.  This was properly delivered into staging first via https://github.com/OpenLiberty/blogs/pull/1258, and then published via the regular staging->prod flow
